### PR TITLE
Div flattening rule

### PR DESCRIPTION
**Pull request by niklasdewally**
Based off #384.

---

Add support for nested divisions, as demonstrated in div_undefzero-03-nested.

Replace the existing div_flattening rule with a more general one, useful for all binary operators.

The flattening rule turns:

c `op1` (a `op2` b) ~> c `op1` x, x = a `op2` b .. where a and b are constants or variables.

When a and b are not constants or variables, I expect flattening tohappen recursively bottom up.

This rule currently only applies to division, as this is our only binary operator; however, it is general enough to be extended to minus,  product, power, etc. when implemented.

This generalization paves the way for adding more arithmetic constraints in a future PR.

### CODE REVIEW

File: conjure_oxide/tests/integration/minion_constraints/div_undefzero-03-nested/input.expected-rewrite.serialised.json
Comment (niklasdewally) [inline conjure_oxide/tests/integration/minion_constraints/div_undefzero-03-nested/input.expected-rewrite.serialised.json:1]: We are introducing lots of identical aux vars here by bubbling before flattening, but from the solutions I think that's
not wrong, just inefficient.

I imagine CSE or some sort of variable merging optimisation would fix this down the line?

Would appreciate your thoughts @ozgurakgun

Comment (niklasdewally) [reply conjure_oxide/tests/integration/minion_constraints/div_undefzero-03-nested/input.expected-rewrite.serialised.json:1]: I've added a nested with zero test and this now doesn't work :(

Comment (niklasdewally) [reply conjure_oxide/tests/integration/minion_constraints/div_undefzero-03-nested/input.expected-rewrite.serialised.json:1]: The problem was that i was flattening SafeDiv, not DivEq.

**Comment by niklasdewally** [comment]
Before I go further with this, I'm going to make a hacky "check against savile row" thing, as I keep missing incorrect solutions.

**Comment by ozgurakgun** [comment]
> Before I go further with this, I'm going to make a hacky "check against savile row" thing, as I keep missing incorrect solutions.

If you are working on this, maybe do a check against `conjure solve --output-format=json --solutions-in-one-file` thing instead? Will be easier to consume the solutions.

**Comment by niklasdewally** [comment]
> solutions

Thank you - I was just looking for the right flags.

I was messing with savilerow -param-to-json, but doing it through conjure would be easier.

**Comment by ozgurakgun** [comment]
You also want `--number-of-solutions=all` probably. And possibly `--copy-solutions=no`, and read the solutions from `conjure-output/*.solutions.json`

**Pull request closed without merging**